### PR TITLE
Disable Plugins menu if no plugins are installed (#2318)

### LINKED ISSUE #2318: If Windows->Plugins menu is empty gray it out
**Is your feature request related to a problem? Please describe.**

The `Windows -> Plugins` should list installed plugins and their related items (if the plugin added something to it). By default, it is empty and even with r2ghidra and r2dec installed it is still empty.
When it is empty, it makes the user confused to think it isn't by enabling the item and showing the Submenu arrow:

image

**Describe the solution you'd like**

If it is empty, disable the menu item but putting its `enabled` member to `false`.
I suggest adding a tooltip that says "No plugins are installed. Check the Plugins section on Cutter documentation to learn more."

**Describe alternatives you've considered**

Remove the menu entirely. I don't recommend this one because then people wouldn't think about this item and Plugin Developers might miss this as well (even though mentioned in the Plugins Dev doc)

### PR DESCRIPTION
**Pull request by sagcan**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://cutter.re/docs/contributing/code/getting-started.html) to this repository
- [x] I made sure to follow the project's [coding style](https://cutter.re/docs/contributing/code/development-guidelines.html)
- [x] I've updated the [documentation](https://cutter.re/docs/user-docs.html) with the relevant information (if needed)

**Detailed description**
The `Windows -> Plugins` entry is now disabled by default and only enabled if there are plugins available to load from (see this [issue](https://github.com/radareorg/cutter/issues/2318) for more information). Note that I was unable to set a Tooltip, not sure if that's how you do it in QT (being new to it) but here's the code I implemented and then deleted after I had no success with it:
```C++
      ui->menuPlugins->setToolTip(QString("tooltip test"));
      ui->menuPlugins->setToolTipsVisible(true);
```

**Test plan (required)**
a) Open Cutter __with__ plugins loaded and check if `Windows -> Plugins` is enabled
b) Open Cutter __without__ any plugins loaded and check if `Windows -> Plugins` is disabled

before_after

**Closing issues**
closes #2318

### CODE REVIEW

**Comment by karliss** [comment]
For the tooltip you might try setting it  on the action which has plugin menu as submenu instead of menu itself. Not sure if it will work, but it's worth trying.

**Comment by sagcan** [comment]
> For the tooltip you might try setting it on the action which has plugin menu as submenu instead of menu itself. Not sure if it will work, but it's worth trying.

Sorry, I don't quite understand what you mean here, could you elaborate? Anyways I've read a bit of QT documentation and I don't think it's possible to have a tooltip on a QMenu object. `setToolTipsVisible` also only enables tooltips for it's actions (i.e. the plugins; my bad) so unless I've missed something we could do the following instead
* add a dummy entry to the Plugins menu with a title like "No plugins loaded, see documentation" or
* change the title of the menu from Plugins to "No plugins loaded" (for which I would then revert the changes inside `MainWindow.ui` and just create an if-else statement instead)

**Comment by karliss** [comment]
Yes `setToolTipsVisible` enables tooltips for actions inside menu so you need to set it for `ui->menuWindows` instead of `ui->menuPlugins` since you are hovering mouse over item inside `menuWindows`. Other aspect is that you need to set the tooltip for action which has menuPlugin assigned as submenu not the menu itself. All the items in a menu are actions including ones that have submenu.

One more detail is that having windows/plugin menu empty is not the same as having no plugins. One plugin can have multiple dockable windows and the opposite, there can be plugins that don't haven any because they use different kind of UI.  @ITAYC0HEN did you considered this?

**Comment by ITAYC0HEN** [comment]
Of course, as can be seen in my issue title and description (notice the parenthesis):




I noticed that the PR is. only checking if the menu is empty, as it should :)

**Comment by karliss** [comment]
> Of course, as can be seen in my issue title and description

I was focusing on the "Describe the solution you'd like" which somewhat contradicts the tittle. In the solution part you wrote "tooltip that says No plugins are installed ...". Having message that no plugins installed when there are plugins, but no plugins with windows would be misleading. That's why I wanted to clarify it.

> I noticed that the PR is. only checking if the menu is empty, as it should :)

No it is checking if the plugin list is empty not if the menu is empty.

**Comment by ITAYC0HEN** [comment]
Sorry, completely my mistakes - twice.


1. Yeah, the tooltip should say that no plugin added items to this menu. We can consider also having two variations. If no plugins are installed, write this. And if they do but none were added actions to this menu then mention this.

2. Need to check if the menu is empty and not the list of installed plugins. Truth is, that if the suggestion above will be accepted, then both things will be checked as there will be two variations of the shown tooltip.


what do you think of having two variations?

**Comment by sagcan** [comment]
@ITAYC0HEN Yup, sounds good to me. I'm just struggling a bit with the tooltip on the QMenu, since it will only show up for it's actions. I've tried implementing @karliss suggestion but that didn't seem to work either and neither did it with `connect`. Are you guys certain that tooltips can be shown on a specific QMenu? Being new to the Qt Framework I, sadly, couldn't find anything.

**Comment by ITAYC0HEN** [comment]
I'd say that it's better not to fight with it too much. From a quick google it doesn't look trivial. Commit the changes without it and we can discuss how to move forward with the tooltip. "worst" case scenario we will give up the idea of using tooltip there.

If @karliss can fetch and try, maybe he'll find a way

**Comment by karliss** [comment]
Yes it is possible and I confirmed it works the way I described before writing my comment above.

**Comment by sagcan** [comment]
With the help of the people over at the Qt Forums I managed to finally display the tooltips (and kinda fully understood what you meant karliss). Anyways hit me up if any changes are necessary, other than that I think the pull request should be complete?

**Review by ITAYC0HEN — APPROVED** [review]
Love it! Good work. Elegant solution :)

File: src/core/MainWindow.cpp [hunk 1]
```diff
@@ -236,8 +236,23 @@ void MainWindow::initUI()
...
+
+    // Check if plugins are loaded and display tooltips accordingly
+    ui->menuWindows->setToolTipsVisible(true);
+    if (plugins.empty()) {
+        ui->menuPlugins->menuAction()->setToolTip(
+                "No plugins are installed. Check the plugins section on Cutter documentation to learn more.");
+        ui->menuPlugins->setEnabled(false);
+    } else if (ui->menuPlugins->isEmpty()) {
+        ui->menuPlugins->menuAction()->setToolTip(
+                "Plugins are installed. No entries haven been added to this menu however.");
```
Comment (ITAYC0HEN) [inline src/core/MainWindow.cpp:254]:
```suggestion
                "The installed plugins didn't add entries to this menu.");
```

Comment (ITAYC0HEN) [reply src/core/MainWindow.cpp:254]: I think it will be better like this, wdyt?

File: src/core/MainWindow.cpp [hunk 2]
```diff
@@ -236,8 +236,23 @@ void MainWindow::initUI()
...
+        for (auto &plugin : plugins) {
+            plugin->setupInterface(this);
+        }
+    }
+
+    // Check if plugins are loaded and display tooltips accordingly
+    ui->menuWindows->setToolTipsVisible(true);
+    if (plugins.empty()) {
+        ui->menuPlugins->menuAction()->setToolTip(
+                "No plugins are installed. Check the plugins section on Cutter documentation to learn more.");
```
Comment (karliss) [inline src/core/MainWindow.cpp:250]: Should use `tr` to make the strings UI strings translatable. https://doc.qt.io/qt-5/i18n-source-translation.html

File: src/core/MainWindow.cpp (hunk 1 above)
Comment (volsa) [reply src/core/MainWindow.cpp]: Sounds good :)

File: src/core/MainWindow.cpp [hunk 3]
```diff
@@ -236,8 +236,23 @@ void MainWindow::initUI()
     connect(ui->actionDocumentation, &QAction::triggered, this, &MainWindow::documentationClicked);
 
     /* Setup plugins interfaces */
-    for (auto &plugin : Plugins()->getPlugins()) {
-        plugin->setupInterface(this);
+    const auto &plugins = Plugins()->getPlugins();
+    if (!plugins.empty()) {
```
Comment (karliss) [inline src/core/MainWindow.cpp:240]: It isn't necessary to do `if (!plugins.empty())` check here now that you made the two tooltip checks bellow.

Comment (sagcan) [reply src/core/MainWindow.cpp]: I missed that one, fixed. Thanks!

**Review by karliss — APPROVED** [review]

**Comment by ITAYC0HEN** [comment]
Good work @sagcan! 
Thank you! Let us know of any other things you would like to work on and try 😊

**Pull request merged**